How to use PSRAM with ESP32 on Zephyr (#351)
espressif/developer-portal: content/blog/2024/12/zephyr-how-to-use-psram/index.md

---
title: "Enabling External PSRAM for Zephyr Applications on ESP32 SoCs"
date: 2024-12-04T00:00:00-00:00
showAuthor: false
authors:
  - "marcio-ribeiro"
tags: ["ESP32", "ESP32-S2", "ESP32-S3", "Zephyr", "PSRAM", "SPIRAM"]
---

Although ESP32 SoCs typically feature a few hundred kilobytes of internal SRAM, some embedded IoT applications—such as those featuring sophisticated graphical user interfaces, handling large data volumes, or performing complex data processing—may require much more RAM. To meet these demands, Espressif provides modules that feature external PSRAM.

> In this article, you will learn how to configure Zephyr to enable your application take full advantage of the PSRAM available on ESP32 modules. We will explore three different strategies for utilizing this PSRAM: using it as a heap for dynamic memory allocation, integrating it as a section into the memory map, and moving code from flash to PSRAM.

Reviewer: to enable your application `to` take...

## Introduction

Before we start, let’s first understand exactly what PSRAM stands for and what it really is. The acronym PSRAM stands for pseudo-static random-access memory. It is a type of dynamic random-access memory (DRAM) that internally manages memory cell refresh, allowing it to function similarly to static random-access memory (SRAM) while retaining the cost and characteristics of DRAM. PSRAM is typically accessible via high-speed serial buses, and some ESP32 series SoCs can communicate with PSRAM and map its contents into the CPU's virtual memory space through a memory management unit (MMU).

In a simplified way, we can describe the relationship among PSRAM, the MMU, and the CPU as follows:

- Data is automatically copied from PSRAM to the data cache when the CPU attempts to access a memory position not currently in the cache.

- Conversely, when CPU modifies data in the cache, the updated contents are copied back to PSRAM.

> As of the time of writing, although ESP32, ESP32-S2, ESP32-S3, ESP32-C61, and ESP32-P4 support external PSRAM, but ESP32-C61 and ESP32-P4 are not currently supported on Zephyr yet.

Reviewer: I found the phrase a bit confusing, specially the word `although`
Maybe rephrase to become clear which HW supports PSRAM and which devices are supported in Zephyr?

[Here](https://www.espressif.com/sites/default/files/documentation/esp32-s3_datasheet_en.pdf#page=13) you can find more details about ESP32-S3 SoC nomenclature.

## Custom hardware and PSRAM

Although Espressif offers modules featuring PSRAM across various ESP32 SoC series, there may be cases where custom hardware must be developed to meet specific requirements, such as PSRAM capacity or the SPI interface bus width, which impacts the time for each memory operation and overall data access speed. Additionally, there may be other reasons unrelated to PSRAM that necessitate custom hardware.

In this case, PSRAM will share data I/O and clock lines with external flash memory, but not chip-select line. PSRAM and flash have separate chip select lines (for example, see the [ESP32-S3 Chip Datasheet](https://www.espressif.com/sites/default/files/documentation/esp32-s3_datasheet_en.pdf#cd-pins-mapping-flash)). For information related on how to connect off-package PSRAM lines, the respective SoC datasheet must be consulted.

The following SPI modes are supported to access external PSRAM memories:

 - Single SPI
 - Dual SPI
 - Quad SPI
 - QPI
 - Octo SPI
 - OPI

> The are limits regarding the maximum PSRAM capacities that each ESP32 SoC series can manage. As an example, ESP32-S3 can manage PSRAMs up to 32 MB.

Reviewer: The ->There
capacities -> capacity

***ESP_SPIRAM_HEAP_SIZE:*** This configuration parameter specifies SPIRAM heap size.

***ESP_SPIRAM_MEMTEST:*** This configuration parameter controls SPIRAM memory test during initialization. It is enabled by default and can be disabled for faster startup.

***SPIRAM_MODE:*** This configuration parameter selects the mode of SPI RAM chip in use. The permitted values are `SPIRAM_MODE_QUAD` and `SPIRAM_MODE_OCT`. Please note that SPIRAM_MODE_OCT is only available in ESP32-S3.

***SPIRAM_TYPE:*** This configuration parameter defines the type of SPIRAM chip in use:

| SPIRAM type | Your PSRAM capacity |
| --- | --- |
| `SPIRAM_TYPE_ESPPSRAM16` | 2 MB |
| `SPIRAM_TYPE_ESPPSRAM32` | 4 MB |
| `SPIRAM_TYPE_ESPPSRAM64` | 8 MB |

***SPIRAM_SPEED:*** This configuration parameter sets the SPIRAM clock speed in MHz:
 | Value | Clock speed |
 | --- | --- |
 | SPIRAM_SPEED_20M  |  20 MHz  |
 | SPIRAM_SPEED_26M  |  26 MHz  |
 | SPIRAM_SPEED_40M  |  40 MHz  |
 | SPIRAM_SPEED_80M  |  80 MHz  |
 | SPIRAM_SPEED_120M | 120 MHz  |

> ***SPIRAM_FETCH_INSTRUCTIONS:*** This configuration parameter allows moving Instructions from flash to PSRAM. If enabled, instructions in flash will be moved into PSRAM on startup. If ```SPIRAM_RODATA``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation does not need to be placed in IRAM, thus optimizing RAM usage. By default, this parameter is disabled.

Reviewer: Instructions -> instructions (maybe no cap)

> ***SPIRAM_RODATA:*** This configuration parameter allows moving read-only data from flash to PSRAM. If ```SPIRAM_FETCH_INSTRUCTIONS``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation  does not need to be placed in IRAM, thus optimizing RAM usage.

Reviewer: operation  does -> operation does (double spacing)

## Installing Zephyr and its dependencies

> To install Zephyr RTOS and the necessary tools, follow the instructions in the Zephyr [Getting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/index.html). By the end of the process, you will have a command-line Zephyr development environment set up and ready to build your application with `west` — the meta-tool responsible for building your application and flashing the generated binary, and other tasks.

Reviewer: instructions in the Zephyr -> in Zephyr's (suggestion only)

Additionally, you need to execute the following command to prepare your environment for building applications for Espressif SoCs:

```sh
west blobs fetch hal_espressif
```

## Dynamically Allocating PSRAM Memory

> PSRAM memory blocks can be made available to applications through Zephyr's shared multi-heap library.  The shared multi-heap memory pool manager uses the multi-heap allocator to manage a set of reserved memory regions with varying capabilities and attributes. For PSRAM, enabling the ```ESP_SPIRAM``` and ```SHARED_MULTI_HEAP``` parameters causes the external PSRAM to be mapped into the data virtual memory space during Zephyr's early initialization stage. The shared multi-heap framework is initialized, and the PSRAM memory region is added to the pool.

Reviewer: library.  The (double spacing)
maybe use the editor tool for a search, review and replace is an easy fix
there are more cases below

CONFIG_SHARED_MULTI_HEAP=y
CONFIG_SPIRAM_MODE_OCT=y
CONFIG_SPIRAM_TYPE_ESPPSRAM64=y
CONFIG_SPIRAM_SPEED_80M=y
CONFIG_ESP32S3_DATA_CACHE_64KB=y
CONFIG_ESP_SPIRAM_MEMTEST=y
```
main.c:

```c
#include <zephyr/kernel.h>
#include <soc/soc_memory_layout.h>
#include <zephyr/multi_heap/shared_multi_heap.h>
#include <zephyr/logging/log.h>

LOG_MODULE_REGISTER(PSRAM_SAMPLE, LOG_LEVEL_INF);

int main(void)
{
	uint32_t *p_mem, k;

	LOG_INF("Sample started");

	p_mem = shared_multi_heap_aligned_alloc(SMH_REG_ATTR_EXTERNAL, 32, 1024*sizeof(uint32_t));

	if (p_mem == NULL) {
		LOG_ERR("PSRAM memory allocation failed!");
		return -ENOMEM;
	}

	for (k = 0; k < 1024; k++) {
		p_mem[k] = k;
	}

	for (k = 0; k < 1024; k++) {
		if (p_mem[k] != k) {
			LOG_ERR("p_mem[%"PRIu32"]: %"PRIu32" (expected value %"PRIu32")", k, p_mem[k], k);
			break;
		}
	}

	shared_multi_heap_free(p_mem);

	if (k < 1024) {
		LOG_ERR("Failed checking memory contents.");
		return -1;
	}

	LOG_INF("Sample finished successfully!");

	return 0;
}
```

To build a project containing these two files targeting a `ESP32-S3-DevKitC-1` board:

```sh
west build -b esp32s3_devkitc/esp32s3/procpu <project folder path> --pristine
```

> This command will create a directory called `build`, which will contain the binary file of our sample along with other intermediate files produced during the building process

Reviewer: add period `.` at the end of phrase

I (184) octal_psram: vendor id    : 0x0d (AP)
I (188) octal_psram: dev id       : 0x02 (generation 3)
I (193) octal_psram: density      : 0x03 (64 Mbit)
I (198) octal_psram: good-die     : 0x01 (Pass)
I (202) octal_psram: Latency      : 0x01 (Fixed)
I (206) octal_psram: VCC          : 0x01 (3V)
I (210) octal_psram: SRF          : 0x01 (Fast Refresh)
I (215) octal_psram: BurstType    : 0x01 (Hybrid Wrap)
I (220) octal_psram: BurstLen     : 0x01 (32 Byte)
I (225) octal_psram: Readlatency  : 0x02 (10 cycles@Fixed)
I (230) octal_psram: DriveStrength: 0x00 (1/1)
I (235) MSPI Timing: PSRAM timing tuning index: 6
I (239) esp_psram: Found 8MB PSRAM device
I (242) esp_psram: Speed: 80MHz
I (651) esp_psram: SPI SRAM memory test OK
I (708) heap_runtime: ESP heap runtime init at 0x3fcacae0 size 243 kB.

*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
[00:00:00.708,000] <inf> PSRAM_SAMPLE: Sample started
[00:00:00.708,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
```

Once the sample finishes executing successfully, we can conclude that the memory allocated from PSRAM was read and written to without any issues.

## Integrating PSRAM memory as a section into the memory map

> Once ```ESP_SPIRAM``` is enabled, a section called `.ext_ram.bss` will be created. This section will hold non-initialized global variables that will later be placed in PSRAM. These gloabal variables must be declared with `__attribute__ ((section (".ext_ram.bss"))`.

Reviewer: gloabal -> global

## 1. Introduction

Before we start, let’s first understand exactly what PSRAM stands for and what it really is. The acronym PSRAM stands for pseudo-static random-access memory. It is a type of dynamic random-access memory (DRAM) that internally manages memory cell refresh, allowing it to function similarly to static random-access memory (SRAM) while retaining the cost and characteristics of DRAM. PSRAM is typically accessible via high-speed serial buses, and some ESP32 series SoCs can communicate with PSRAM and map its contents into the CPU's virtual memory space through a memory management unit (MMU).

In a simplified way, we can describe the relationship among PSRAM, the MMU, and the CPU as follows:

- Data is automatically copied from PSRAM to the data cache when the CPU attempts to access a memory position not currently in the cache.

- Conversely, when CPU modifies data in the cache, the updated contents are copied back to PSRAM.

> As of the time of writing, although ESP32, ESP32-S2, ESP32-S3, ESP32-C61, and ESP32-P4 support external PSRAM, ESP32-C6 and ESP32-P4 are not currently supported on Zephyr yet.

Author: ESP32-C61

I (193) octal_psram: density      : 0x03 (64 Mbit)
I (198) octal_psram: good-die     : 0x01 (Pass)
I (202) octal_psram: Latency      : 0x01 (Fixed)
I (206) octal_psram: VCC          : 0x01 (3V)
I (210) octal_psram: SRF          : 0x01 (Fast Refresh)
I (215) octal_psram: BurstType    : 0x01 (Hybrid Wrap)
I (220) octal_psram: BurstLen     : 0x01 (32 Byte)
I (225) octal_psram: Readlatency  : 0x02 (10 cycles@Fixed)
I (230) octal_psram: DriveStrength: 0x00 (1/1)
I (235) MSPI Timing: PSRAM timing tuning index: 6
I (238) esp_psram: Found 8MB PSRAM device
I (242) esp_psram: Speed: 80MHz
I (301) heap_runtime: ESP heap runtime init at 0x3fcacad0 size 243 kB.

*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
[00:00:00.301,000] <inf> PSRAM_SAMPLE: Sample started
[00:00:00.373,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
```

Here, we can observe that the application now starts earlier compared to the previous sample. This improvement results from disabling the **CONFIG_ESP_SPIRAM_MEMTEST** parameter, which bypasses the PSRAM memory test that previously took a few hundred milliseconds.

## 8. Moving code from FLASH to PSRAM

> Enabling **CONFIG_SPIRAM_FETCH_INSTRUCTIONS** will move instructions from FLASH to PSRAM during startup, and if **CONFIG_SPIRAM_RODATA** is also enabled, code that requires execution during an SPI1 Flash operation can forgo being placed in IRAM, thus optimizing RAM usage.

Author: I'll try to address this matter keeping the "parameter name" along the explanations and use `CONFIG_` in sample code as user will have to do.

I (235) MSPI Timing: PSRAM timing tuning index: 6
I (238) esp_psram: Found 8MB PSRAM device
I (242) esp_psram: Speed: 80MHz
I (301) heap_runtime: ESP heap runtime init at 0x3fcacad0 size 243 kB.

*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
[00:00:00.301,000] <inf> PSRAM_SAMPLE: Sample started
[00:00:00.373,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
```

Here, we can observe that the application now starts earlier compared to the previous sample. This improvement results from disabling the ```ESP_SPIRAM_MEMTEST``` parameter, which bypasses the PSRAM memory test that previously took a few hundred milliseconds.

## Running Code from PSRAM Instead of FLASH

Enabling the ```SPIRAM_FETCH_INSTRUCTIONS``` parameter will move instructions from FLASH to PSRAM during startup, and if the ```ESP_SPIRAM_MEMTEST``` parameter is also enabled, code that requires execution during an SPI1 Flash operation can forgo being placed in IRAM, thus optimizing RAM usage.

To check the effects of ```SPIRAM_FETCH_INSTRUCTIONS``` and ```SPIRAM_RODATA``` parameters, let's build `hello_world` first without enabling them and then enable them.

> ***Building `hello_world` with `SPIRAM_FETCH_INSTRUCTIONS` and `SPIRAM_RODATA` disbled***

Reviewer 2: ```suggestion
***Building `hello_world` with `SPIRAM_FETCH_INSTRUCTIONS` and `SPIRAM_RODATA` disabled***
```

I (252) mmu_psram: Read only data copied and mapped to SPIRAM
I (652) esp_psram: SPI SRAM memory test OK
I (708) heap_runtime: ESP heap runtime init at 0x3fc93030 size 345 kB.

*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
Hello World! esp32s3_devkitc/esp32s3/procpu
```

Here, we can confirm that instructions and read-only data were copied and mapped to PSRAM, which can optimize RAM usage when there is code that needs to execute during an SPI1 flash operation.

## Letting ESP32 Wi-Fi Use PSRAM

Another potential use of PSRAM to reduce SRAM usage is by enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` parameter, allowing the ESP32 Wi-Fi stack to dynamically allocate memory from PSRAM.

To check the results, let's build the `zephyr/samples/net/wifi/shell`, first with `ESP32_WIFI_NET_ALLOC_SPIRAM` disabled and then with this parameter enabled. In both cases let's flash the binary into the board and launch a console to interact with the shell. After connect to an access point we will execute `net allocs` to get information from where memory is being allocated in both cases.

***Building `zephyr/samples/net/wifi/shell` with `ESP32_WIFI_NET_ALLOC_SPIRAM` disabled***

> Building, flashing and monitoring:

Reviewer 2: ```suggestion
Building, flashing, and monitoring:
```

I (686) esp_psram: SPI SRAM memory test OK
I (714) heap_runtime: ESP heap runtime init at 0x3fca8788 size 259 kB.

*** Booting Zephyr OS build v4.0.0-2029-g6e0e5591a4fb ***
uart:~$ wifi scan
Scan requested

Num  | SSID                             (len) | Chan (Band)   | RSSI | Security        | BSSID             | MFP
1    | Soares                           6     | 11   (2.4GHz) | -63  | WPA2-PSK        | 90:0A:62:42:A5:BF | Disable
...
uart:~$ wifi connect --key-mgmt 1 --ssid <SSID> --passphrase <PASSPHRASE>
Connection requested
Connected
[00:01:31.902,000] <inf> net_dhcpv4: Received: 192.168.15.2
uart:~$ net ping 192.168.15.17
PING 192.168.15.17
28 bytes from 192.168.15.17 to 192.168.15.2: icmp_seq=1 ttl=64 time=231 ms
28 bytes from 192.168.15.17 to 192.168.15.2: icmp_seq=2 ttl=64 time=6 ms
28 bytes from 192.168.15.17 to 192.168.15.2: icmp_seq=3 ttl=64 time=338 ms
uart:~$ net allocs
Network memory allocations
memory		Status	Pool	Function alloc -> freed
0x3c083920	 free	   RX	eth_esp32_rx():135 -> processing_data():178
0x3c0836a0	 free	   TX	dhcpv4_create_message():287 -> ethernet_send():762
0x3c082894	 free	TDATA	ethernet_fill_header():532 -> ethernet_remove_l2_header():61
```

> Examining the output from both sessions, we can confirm that in the first case, network stack allocations were made using memory from internal RAM -- addresses 0x3fca4af0, 0x3fca4870, and 0x3fca3a2c. We can also confirm that in the second case, network stack allocations were taking memory from PSRAM -- addresses 0x3c083920, 0x3c0836a0, and 0x3c082894 -- thus proving that enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` avoids allocating memory from the precious internal SRAM and takes advantage of PSRAM.

Reviewer 2: ```suggestion
Examining the output from both sessions, we can confirm that in the first case, network stack allocations were made using memory from internal RAM -- addresses 0x3fca4af0, 0x3fca4870, and 0x3fca3a2c. We can also confirm that in the second case, network stack allocations were taking memory from PSRAM -- addresses `0x3c083920`, `0x3c0836a0`, and `0x3c082894` -- thus proving that enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` avoids allocating memory from the precious internal SRAM and takes advantage of PSRAM.
```

## Final Thoughts

> Throughout this article, we explored three different strategies for utilizing PSRAM: using it as a heap for dynamic memory allocation, integrating it as a section in the memory map, and moving code from flash to PSRAM.

Reviewer 2: In the intro and here, you talk about the three strategies as this:

- using it as a heap for dynamic memory allocation
- integrating it as a section in the memory map
- moving code from flash to PSRAM

However, the actual article sections are called:

- Using PSRAM for Dynamic Memory Allocation
- Adding PSRAM to Your ESP32 Memory Map
- Running Code from PSRAM Instead of Flash

Suggest that you use the same names for these strategies both in text and section titles.

> In addition to these three strategies, it is also possible to leverage PSRAM to execute external code compiled as Position Independent Code (PIC), which can be placed in PSRAM by a loader for execution. You can find more information regarding to this way of taking avantage of PSRAM consulting Zephyr's documentation about [LLEXT -- Linkable Loadable Extensions](https://docs.zephyrproject.org/latest/services/llext/index.html)

Reviewer 2: ```suggestion
In addition to these three strategies, it is also possible to leverage PSRAM to execute external code compiled as Position Independent Code (PIC), which can be placed in PSRAM by a loader for execution. You can find more information about this way of using PSRAM by consulting Zephyr's document [LLEXT -- Linkable Loadable Extensions](https://docs.zephyrproject.org/latest/services/llext/index.html).
```
